### PR TITLE
506 Constructed Payloads

### PR DESCRIPTION
**Pull request by askirk**
- See Issue #506

### CODE REVIEW

File: sdl_android/src/main/java/com/smartdevicelink/protocol/SdlPacket.java [hunk 1]
```diff
@@ -65,6 +67,7 @@
 	int messageId;
 	int priorityCoefficient;
 	byte[] payload = null;
+	HashMap<String, Object> bsonPayload = new HashMap<>();
```
Comment (joeygrover) [inline sdl_android/src/main/java/com/smartdevicelink/protocol/SdlPacket.java]: We shouldn't instantiate this unless it is needed. The majority of packets won't use it so it's best to not allocate the memory.

File: sdl_android/src/main/java/com/smartdevicelink/protocol/SdlPacket.java [hunk 2]
```diff
@@ -186,6 +189,11 @@ public long getDataSize() {
 	}
 	
 	public byte[] constructPacket(){
+		if(!bsonPayload.isEmpty()){
```
Comment (joeygrover) [inline sdl_android/src/main/java/com/smartdevicelink/protocol/SdlPacket.java]: After changing how this is instantiated, we will also have to check for null

File: sdl_android/src/main/java/com/smartdevicelink/protocol/SdlPacket.java [hunk 3]
```diff
@@ -344,6 +353,22 @@ public SdlPacket createFromParcel(Parcel in) {
 		}
 
     };
-	
-	
+
+	public void putTag(String tag, Object data){
+		bsonPayload.put(tag, data);
```
Comment (joeygrover) [inline sdl_android/src/main/java/com/smartdevicelink/protocol/SdlPacket.java]: This is where you should instantiate the `bsonPayload` object `if(bsonPayload==null){bsonPayload = new HashMap<String,Object>());}`

File: sdl_android/src/main/java/com/smartdevicelink/protocol/SdlPacket.java [hunk 4]
```diff
@@ -344,6 +353,22 @@ public SdlPacket createFromParcel(Parcel in) {
...
+	}
+
+	public Object getTag(String tag){
+		if(payload == null){
+			return null;
+		}else if(bsonPayload.isEmpty()){
```
Comment (joeygrover) [inline sdl_android/src/main/java/com/smartdevicelink/protocol/SdlPacket.java]: Again will have to add a null check

File: sdl_android/src/main/java/com/smartdevicelink/protocol/enums/BsonTags.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package com.smartdevicelink.protocol.enums;
+
+public class BsonTags {
```
Comment (joeygrover) [inline sdl_android/src/main/java/com/smartdevicelink/protocol/enums/BsonTags.java]: We should make this a little better organized; creating subclasses in this class would be fine. Open to suggestions. Also adding descriptions to the constants.

File: sdl_android/src/main/java/com/smartdevicelink/protocol/WiProProtocol.java [hunk 6]
```diff
@@ -1,30 +1,39 @@
 package com.smartdevicelink.protocol;
 
+
+import android.util.Log;
```
Comment (joeygrover) [inline sdl_android/src/main/java/com/smartdevicelink/protocol/WiProProtocol.java]: Should remove this import if it's not used.

Comment (askirk) [inline sdl_android/src/main/java/com/smartdevicelink/protocol/WiProProtocol.java]: was going to use it for printing out rejected Params as error log

Comment (joeygrover) [inline sdl_android/src/main/java/com/smartdevicelink/protocol/WiProProtocol.java]: I missed that. We should pass those params back. I don't think they are helpful to long unless a developer is debugging. Maybe for now just leave a //TODO and with the modifications for the video streaming params we can pass back the rejected params.

**Comment by askirk** [comment]
@joeygrover updated

**Pull request merged**